### ISSUE TITLE
Interceptor class methods do not support nullable return types

### ISSUE THREAD

**Issue by jsamhall** [magento/magento2#15505]
### Preconditions

1. Magento 2.3.1-develop (cloned from magento-engcom/php-7.2-support)
2. PHP 7.2.5.1

### Steps to reproduce

1. Declare a public method with a nullable return type on a class that will have an Interceptor generated for it. In this case, the class extends from the Session Manager:
```
namespace MyVendor\MyModule\Model;

class MySession extends Magento\Framework\Session\SessionManager
{
    public function getSomeString() : ?string
    {
           return $this->getData('something');
    }
}
```

2. Trigger code generation in your environment.

### Expected result

1. The method signature on the Interceptor should be the exact same as the overridden method signature.

### Actual result

1. The method signature on the Interceptor does not include the nullable return type.
```
namespace MyVendor\MyModule\Model\MySession;

class Interceptor extends \MyVendor\MyModule\Model\Session
{
    public function getSomeString() : string
    {
        $pluginInfo = $this->pluginList->getNext($this->subjectType, 'getSomeString');
        if (!$pluginInfo) {
            return parent::getSomeString();
        } else {
            return $this->___callPlugins('getSomeString', func_get_args(), $pluginInfo);
        }
    }
}
```

### Further Investigation
After digging into this issue, I'm not sure if this is Magento or Zend's fault. It is clear via [this declined RFC ](https://wiki.php.net/rfc/reflectiontypeimprovements) that `ReflectionNamedType` probably should have included the `?` nullable indicator when the object is represented as a `String` and `allowsNull` is `true`, however this RFC was declined, therefore this change did not occur. To me, this feels like an oversight in PHP, and should be better documented. I digress; onto the investigation. 

During the code generation procedure for the Interceptor, when adding methods (in `Magento/Framework/Code/Generator/ClassGenerator::addMethods`), Magento creates a new instance of Zend's `MethodGenerator` class and proceeds to "manually" build the object with setters. This approach is an alternative to using Zend's `MethodGenerator::fromReflection`, which calls for a \ReflectionMethod. Presumably this method is not used because Magento's `Code\Generator\Interceptor` class `_getClassMethods` returns an array of method data (strings, objects, Reflections, etc.) and not an array of \ReflectionMethod object, hence the Zend class' named constructor cannot be used (which does account for a nullable `ReflectionNamedType` if you dig into it).

Jumping back into `Magento/Framework/Code/Generator/ClassGenerator::addMethods` where Magento is building the object with setters, we find this on line 134:

` $methodObject->setReturnType($methodOptions['returnType']);`

Where `$methodObject` is `Zend\Code\Generator\MethodGenerator` and `$methodOptions['returnType']` is a `ReflectionNamedType` that appropriately returns `true` when `allowsNull` is invoked. 

Here is where I'm unsure if it is Zend's mishandling of `ReflectionNamedType` or Magento's improper use of Zend Framework's `MethodGenerator`.

The `setReturnType` method typehints a string (so `ReflectionNamedType` works as it implements `__toString()`) or null, as so:

```
    /**
     * @param string|null
     *
     * @return MethodGenerator
     */
    public function setReturnType($returnType = null)
    {
        $this->returnType = null === $returnType
            ? null
            : TypeGenerator::fromTypeString($returnType);

        return $this;
    }
```

To me, this suggests maybe Magento is misusing Zend's framework by erroneously passing it an instance of `ReflectionNamedType`, though it is equally possible that Zend has overlooked how `ReflectionNamedType` functions (i.e., it is castable to a string). As we dig into `fromTypeString`, we find the following:

```
    public static function fromTypeString($type)
    {
        list($nullable, $trimmedNullable) = self::trimNullable($type);
        list($wasTrimmed, $trimmedType) = self::trimType($trimmedNullable);
```

Remember, `$type` is an instance of `ReflectionNamedType`. As you see, it is passed over to `trimNullable` which does this:

```
    private static function trimNullable($type)
    {
        if (0 === strpos($type, '?')) {
            return [true, substr($type, 1)];
        }

        return [false, $type];
    }
```
This is where I'm unsure if Zend is not accounting for the `ReflectionNamedType` being cast to a string or is unaware of PHP's behavior excluding the `?` nullable indicator when `__toString()` is invoked. Perhaps worth noting, `ReflectionNamedType__toString()` method is marked as deprecated as of PHP 7.1

When a `ReflectionNamedType` is nullable return type (i.e., `allowsNull()` returns true), `getName()` (or casting to a string, as is done here, seemingly intentionally by Magento's team) does NOT include the `?` nullable indicator; rather it only includes the scalar type (e.g. "string"). Therefore, the generated method footprint on the Interceptor class does not include the `?` nullable indicator, and the described issue occurs.

**Comment by jsamhall** [comment]
According to [this declined RFC](https://wiki.php.net/rfc/reflectiontypeimprovements) it appears the exclusion of ? when casting to the object to a string was an oversight/mistake due to lackluster documentation. The RFC proposed fixing this (by ensuring ? is included for nullable types) which classified as a BC break. Therefore I think the bug is Magento (via misuse of Zend) or Zend's (via poor handling of ReflectionNamedType) responsibility.

**Comment by engcom-backlog-nickolas** [comment]
Hello @jsamhall, thank you for your report.
This issue already fixed in current 2.3-develop.

**Comment by molovo** [comment]
@engcom-backlog-nickolas do you know if the fix for this is to be backported to 2.2.x?

**Comment by magento-engcom-team** [comment]
Hi @jsamhall. Thank you for your report. 
The issue has been fixed in magento/magento2#19398 by @swnsma in 2.2-develop branch
Related commit(s):
 - [8b900d4131daedc5daf46d9974d83ff59b02100a](https://github.com/magento/magento2/commit/8b900d4131daedc5daf46d9974d83ff59b02100a)
 - [f674badeadb1032a5b29e152de1503a55049b893](https://github.com/magento/magento2/commit/f674badeadb1032a5b29e152de1503a55049b893)
 - [541ce189de4b51fcb49bc292037e88c050300063](https://github.com/magento/magento2/commit/541ce189de4b51fcb49bc292037e88c050300063)
 - [d33db6438b1c0ef9a2590a565b2d9ad8cc7f9e7d](https://github.com/magento/magento2/commit/d33db6438b1c0ef9a2590a565b2d9ad8cc7f9e7d)
 - [c0c0a37082c5596d3e7cae18caf33c9cf1f4cf27](https://github.com/magento/magento2/commit/c0c0a37082c5596d3e7cae18caf33c9cf1f4cf27)
 - [3e7f0eb296b53b9a5070bee0270664cda2577004](https://github.com/magento/magento2/commit/3e7f0eb296b53b9a5070bee0270664cda2577004)
 - [d7efcca141098e910a0265f9ba79b7c6aac655ae](https://github.com/magento/magento2/commit/d7efcca141098e910a0265f9ba79b7c6aac655ae)
 - [d2490367cb778b4c417c0bbca1bad8909f8f32a9](https://github.com/magento/magento2/commit/d2490367cb778b4c417c0bbca1bad8909f8f32a9)
 - [87ab26f8c2ea079ebdf74d0dd0ba7e3545ccdd13](https://github.com/magento/magento2/commit/87ab26f8c2ea079ebdf74d0dd0ba7e3545ccdd13)
 - [41c8da0d83981e26931c778c95f0b893f28ee006](https://github.com/magento/magento2/commit/41c8da0d83981e26931c778c95f0b893f28ee006)
 - [f0758f325820507e554ca2aa52ffc6d214d1909e](https://github.com/magento/magento2/commit/f0758f325820507e554ca2aa52ffc6d214d1909e)
 - [a33f89d0247633a13234bac4e2d505389b17c245](https://github.com/magento/magento2/commit/a33f89d0247633a13234bac4e2d505389b17c245)

The fix will be available with the upcoming 2.2.8 release.
